packages/webapp/src/components/Form/CompositionInputs/NumberInputWithSelect.tsx

```diff
@@ -0,0 +1,119 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+import { ReactElement } from 'react';
+import clsx from 'clsx';
+import { useTranslation } from 'react-i18next';
+import ReactSelect from '../ReactSelect';
+import useReactSelectStyles from '../Unit/useReactSelectStyles';
+import useNumberInput from '../NumberInput/useNumberInput';
+import InputBase from '../InputBase';
+import { styles as reactSelectDefaultStyles } from '../ReactSelect';
+import styles from './styles.module.scss';
+
+type ReactSelectOption = { value: string; label: ReactElement | string };
+
+export type NumberInputWithSelectProps<T extends string, U extends string> = {
+  name: T;
+  unitName: U;
+  label: string;
+  unitOptions: ReactSelectOption[];
+  disabled?: boolean;
+  error?: string;
+  className?: string;
+  values?: { [K in T | U]?: number | ReactSelectOption | undefined };
+  onChange: (fieldName: T | U, value: number | ReactSelectOption) => void;
+};
+
+const REACT_SELECT_WIDTH = 44;
+
+const NumberInputWithSelect = <T extends string, U extends string>({
+  name,
+  unitName,
+  label,
+  unitOptions,
+  disabled,
+  error,
+  className,
+  onChange,
+  values = {},
+}: NumberInputWithSelectProps<T, U>) => {
+  const { t } = useTranslation();
+  const unit = values[unitName] as ReactSelectOption;
```

Comment:
Can using 'as' be avoided? I'm thinking if values is typed instead of a generic then that could work? Ex: 

```
type CompositionValues = {
    unit: Unit;
    n: number;
    p: number;
    k: number;
}

```
Then in the story it can be used to type the form like useForm<{composition: CompositionValues}>

Reply:
Let me try...